enh(php) support C style and hash prefix single line comments in functions

hashar opened the pull request:
enh(php) add // and # comments in functions and attributes

functions, arrow functions parameters can contain nested comments such as:

    header( /* cache age */ 'Age: 60' );
    fn($x /* something */) => $x;

Those are C blocks comments. However PHP also support single line comments either C-style (`//`) or using hash (`#`):

    header(
        # Set caching header
        'Age: 60' // seconds
    );
    fn(
        # We like magic
        $x // something
    ) => $x;

PHP 8 attributes can also have such comments:

    #[\Deprecated(
    # A comment
    since: "2025-03-25"
    // C line comment
    #)]

Parameters of functions, arrow functions and PHP 8 attributes solely had `C_BLOCK_COMMENT_MODE`.

Add `C_LINE_COMMENT_MODE` and `HASH_COMMENT_MODE` to recognize the single line comments.

Fixes: #3655
Fixes: https://phabricator.wikimedia.org/T372404

hashar commented:
Looks like the single lines PHP comments that starts with `# ` suffer from the same issue (#3655) and this patch should fix them as well while at it.

joshgoebel commented on test/markup/php/functions.txt:
```diff
@@ -49,3 +49,15 @@ setAlarm(
     time:time() + array(5)[0] + Foo::HOUR,
 );
+
+header(
```
Is header a PHP thing?

wkania replied:
In this context (functions.txt), he calling user definer function:

> header()

hashar replied:
> Is header a PHP thing?

Yes! That is used to add a HTTP response header to the output. PHP is pretty much a language to generate web pages and their responses and `header()` is I guess known to anyone practicing PHP.  https://www.php.net/manual/en/function.header.php

The function name should not matter to highlight.js, as long as it is recognized as a function. If `header()` is confusing, I do not mind changing it to another built-in function that is common to other languages such as `printf()`.

joshgoebel reviewed (changes requested):
Please add an entry to CHANGES.md as well!  Looks good!

paladox commented on test/markup/php/functions.txt:
```diff
@@ -49,3 +49,15 @@ setAlarm(
     label: 'foo',
```
Maybe you want to add a test that tests a comment with a apostrophe?

hashar replied:
> Maybe you want to add a test that tests a comment with a apostrophe?

The single quotes or apostrophes are irrelevant to the issue, they were merely a symptom. Since the text was NOT wrapped as a comment, when Highlight.js encountered the single quote it would start a string section which changed the color. That surfaced the root cause: the text was not considered a comment but some inline PHP.

wkania commented on src/languages/php.js:
```diff
@@ -438,6 +440,8 @@ export default function(hljs) {
     LEFT_AND_RIGHT_SIDE_OF_DOUBLE_COLON,
     hljs.C_BLOCK_COMMENT_MODE,
+    // Attributes are single line and can neither contain C_LINE_COMMENT_MODE
+    // nor hljs.HASH_COMMENT_MODE
```
This statement is invalid. Attributes are [multiline](https://onlinephp.io?s=s7EvyCjg5VKOjnFJLShKTU4sSU3RAPIVkvNzc1PzSni5ijPzklOtFJSMDIxMdA1MdQ3MlXi59PURCjRjFXi50krzkksy8_MUSlKLSzQ0Fap5uThTkzPyFdQT1a15uWp5uSAS1gA%2C&v=8.4.5%2C8.2.20)

hashar replied:
Thank you, I have removed the comment. I confess: I have never heard of PHP 8 attributes before :)

wkania commented on src/languages/php.js:
```diff
@@ -438,6 +440,8 @@ export default function(hljs) {
     NAMED_ARGUMENT,
     LEFT_AND_RIGHT_SIDE_OF_DOUBLE_COLON,
     hljs.C_BLOCK_COMMENT_MODE,
+    // Attributes are single line and can neither contain C_LINE_COMMENT_MODE
+    // nor hljs.HASH_COMMENT_MODE
```
```suggestion
    hljs.C_LINE_COMMENT_MODE,
    hljs.HASH_COMMENT_MODE,
```

hashar commented:
I have made adjustments to properly support attributes, thanks to @wkania

> Please add an entry to CHANGES.md as well! Looks good!

I have added an entry in the change log, thank you!

hashar commented:
I have addressed the issues pointed by @wkania .

Looks like this is now pending on @joshgoebel to acknowledge my answer to their remark at https://github.com/highlightjs/highlight.js/pull/4230#pullrequestreview-2708111829 ?
